zebra-consensus/src/primitives/halo2/tests.rs

```diff
@@ -0,0 +1,146 @@
+//! Tests for verifying simple Halo2 proofs with the async verifier
+
+use std::convert::TryInto;
+
+use futures::stream::{FuturesUnordered, StreamExt};
+use tower::ServiceExt;
+
+use halo2::{arithmetic::FieldExt, pasta::pallas};
+use orchard::{
+    builder::Builder,
+    bundle::Flags,
+    circuit::ProvingKey,
+    keys::{FullViewingKey, SpendingKey},
+    value::NoteValue,
+    Anchor, Bundle,
+};
+use rand::rngs::OsRng;
+
+use zebra_chain::orchard::ShieldedData;
+
+use crate::primitives::halo2::*;
+
+lazy_static::lazy_static! {
+    pub static ref PROVING_KEY: ProvingKey = ProvingKey::build();
+}
+
+async fn verify_orchard_halo2_proofs<V>(
+    verifier: &mut V,
+    shielded_data: Vec<ShieldedData>,
+) -> Result<(), V::Error>
+where
+    V: tower::Service<Item, Response = ()>,
+    <V as tower::Service<Item>>::Error: std::convert::From<
+        std::boxed::Box<dyn std::error::Error + std::marker::Send + std::marker::Sync>,
+    >,
+    <V as tower::Service<Item>>::Error: std::fmt::Debug,
+{
+    let mut async_checks = FuturesUnordered::new();
+
+    for sd in shielded_data {
+        tracing::trace!(?sd);
+
+        let rsp = verifier.ready().await?.call(Item::from(sd));
+
+        async_checks.push(rsp);
+    }
+
+    while let Some(result) = async_checks.next().await {
+        tracing::trace!(?result);
+        result?;
+    }
```

Comment:
Suggestion: test if verification fails with a corrupted input or proof, to make sure we're not accidentally always returning success.

Reply:
Done in  ea16b947d69c485e9aa1889fe205281ee4f2e044